documentation/projects/proposals/trust_and_safety/content_report_moderation/20230517-implementation_plan_mature_copy_change.md

```diff
@@ -0,0 +1,294 @@
+`sensitive` and `sensitive_filtered` respectively. The migration necessary for
+changing the data is described below, however the API will need to continue to
+accept `mature` as a report reason and convert it internally to `sensitive` (at
+least until we decide to make a version change to the API). The media report
+tables already use table aliases as well
+([`ImageReport`](https://github.com/WordPress/openverse/blob/7b95a4c8eaa9804f53b4be7ac969e04ca437695a/api/api/models/image.py#L121-L122)
+and
+[`AudioReport`](https://github.com/WordPress/openverse/blob/7b95a4c8eaa9804f53b4be7ac969e04ca437695a/api/api/models/audio.py#L304-L305)) -
+since we're already performing table renames as part of this IP, it makes sense
+to rename these tables for the same reason.
+
+In order to perform both of these changes in a
+[zero-downtime manner](https://docs.openverse.org/general/zero_downtime_database_management.html),
+the following steps will need to be taken:
+
+1. Add new models with updated names and fields for `SensitiveImage` and
+   `SensitiveAudio`. Also add new models for `ImageReport` and `AudioReport`.
+   This will require renaming code references from the existing `ImageReport`
+   and `AudioReport` models to `NsfwReport` and `NsfwReportAudio` temporarily
+   while maintaining the reference to the existing table name. Said another way:
+   - `ImageReport` -> `NsfwReport` _(code only, `db_table` should remain the
+     same)_
+   - `AudioReport` -> `NsfwReportAudio` _(code only, `db_table` should remain
+     the same)_
+   - Add `ImageReport` _(new model, `db_table` will not need to be overridden)_
+   - Add `AudioReport` _(new model, `db_table` will not need to be overridden)_
+2. Modify the API to write to both the old and new tables for each of the above
+   instances. On the new report tables, the `mature_filtered` reason should be
+   written as `sensitive_filtered`. Data should continue to be read from the
+   original tables during this time.
+3. Deploy this version of the API.
+4. Create a
+   [data management command](https://docs.openverse.org/general/zero_downtime_database_management.html#django-management-command-based-data-transformations)
+   which copies all data from the original tables to the new tables. The
+   `identifier` field can be used for determining which rows have already been
+   copied.
+5. Deploy this version of the API and run the data management command until no
+   more rows need to be copied for each of the columns.
+6. Modify the ingestion server to use the new tables as reference for
+   determining mature status. Specifically the
+   [`exists_in_mature` table check](https://github.com/WordPress/openverse/blob/765a24028baa922b27fc0c38b8ae4c69902eec40/ingestion_server/ingestion_server/queries.py#L33-L37)
+   (this will necessarily involve updating tests and initialization SQL as
+   well).
+7. Remove all code references to the old model from the API, but leave the
```

Comment:
Which model exactly are we speaking of here? Is it in the code from the Ingestion server?

Reply:
This refers to the `MatureImage` and `MatureAudio` models, which are replaced by `SensitiveImage` and SensitiveAudio`. I'll make that explicit here. All the steps here deal with the API, unless explicitly stated like in step 6.